src/Stack/Build/Execute.hs

```diff
@@ -1928,10 +1929,11 @@ singleTest topts testsToRun ac ee task installedMap = do
                 -- see e.g. https://github.com/doctest/issues/119
                 -- also we set HASKELL_DIST_DIR to a package dist directory so
                 -- doctest will be able to load modules autogenerated by Cabal
+                ghcEnvRandomId <- liftIO (randomIO :: IO Int)
                 let setEnv f pc = modifyEnvVars pc $ \envVars ->
                       Map.insert "HASKELL_DIST_DIR" (T.pack $ toFilePath buildDir) $
                       Map.insert "GHC_ENVIRONMENT" (T.pack f) envVars
-                    fp = toFilePath $ eeTempDir ee </> testGhcEnvRelFile
+                    fp = toFilePath (eeTempDir ee </> testGhcEnvRelFile) <> show ghcEnvRandomId
```

Comment:
The `ghcEnvRandomId` line interjects the preceding multiline comment explaining env variables and the next line, `let setEnv ...` — to which that comment is relevant. Please move somewhere else.

Perhaps like this?

```diff
                -- see e.g. https://github.com/doctest/issues/119
                -- also we set HASKELL_DIST_DIR to a package dist directory so
                -- doctest will be able to load modules autogenerated by Cabal
                let setEnv f pc = modifyEnvVars pc $ \envVars ->
                      Map.insert "HASKELL_DIST_DIR" (T.pack $ toFilePath buildDir) $
                      Map.insert "GHC_ENVIRONMENT" (T.pack f) envVars
+               ghcEnvRandomId <- liftIO (randomIO :: IO Int)
-                   fp = toFilePath $ eeTempDir ee </> testGhcEnvRelFile
+               let fp = toFilePath (eeTempDir ee </> testGhcEnvRelFile) <> show ghcEnvRandomId
```

Further, it's not so hard to stay within typed [`Path`](https://hackage.haskell.org/package/path), just import [`addExtension`](https://hackage.haskell.org/package/path-0.9.2/docs/Path-Posix.html#v:addExtension):
```diff
-               ghcEnvRandomId <- liftIO (randomIO :: IO Int)
-               let fp = toFilePath $ eeTempDir ee </> testGhcEnvRelFile
+               randomSuffix <- ("-"<>) . show <$> liftIO (randomIO :: IO Int)
+               fp <- toFilePath <$> addExtension randomSuffix (eeTempDir ee </> testGhcEnvRelFile)
```